### PR TITLE
[Tabular] Added use_bag_holdout functionality

### PR DESCRIPTION
**Pull request by Innixma**
*Issue #, if available:*

*Description of changes:*

- Added use_bag_holdout option to allow bagging/stacking with a holdout set. (This is not equivalent to blending)
- Greatly improved handling of rare classes in various situations + bug fixes to rare class duplication logic.

Before merging:

- [x] Add unit test
- [x] Benchmark
 
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-1105-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1105/1/index.html

**Comment by szha** [comment]
Job PR-1105-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1105/2/index.html

**Comment by szha** [comment]
Job PR-1105-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1105/4/index.html

**Comment by szha** [comment]
Job PR-1105-5 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1105/5/index.html

**Comment by szha** [comment]
Job PR-1105-6 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1105/6/index.html

**Comment by Innixma** [comment]
Benchmark Results show slight improvement:

```
                                        framework  > AutoGluon_bestquality_1h_2021_06_08_holdout  < AutoGluon_bestquality_1h_2021_06_08_holdout  = AutoGluon_bestquality_1h_2021_06_08_holdout  % Less Avg. Errors Than AutoGluon_bestquality_1h_2021_06_08_holdout  time_train_s  metric_error  time_infer_s  loss_rescaled  time_train_s_rescaled  time_infer_s_rescaled      rank  rank=1_count  rank=2_count  rank=3_count  rank>3_count  error_count
0     AutoGluon_bestquality_1h_2021_06_08_holdout                                              0                                              0                                            998                                           0.000000                     3363.007315  78895.417262     34.595892       0.442886               1.011143               1.142435  1.484469           462           536             0             0            0
1  AutoGluon_bestquality_1h_2021_06_05_fastai_fix                                            431                                            462                                            105                                          -0.000324                     3365.012926  71445.497229     34.803407       0.474950               1.011680               1.120427  1.515531           431           567             0             0            0
```

These results are without using the `use_bag_holdout` option, and the improvement comes from the improved handling of rare classes in multi-class classification problems.

Benchmarking the `use_bag_holdout` option shows that it is significantly worse on average, however for larger datasets (>=100,000 rows) it is on-par or better. This is because the amount of data held-out is a smaller fraction than in other datasets and the benefits of avoiding information leakage outweigh the slight reduction in training data:

```
                                     framework  > AutoGluon_bestquality_1h_2021_06_08_holdout  < AutoGluon_bestquality_1h_2021_06_08_holdout  = AutoGluon_bestquality_1h_2021_06_08_holdout  % Less Avg. Errors Than AutoGluon_bestquality_1h_2021_06_08_holdout  time_train_s  metric_error  time_infer_s  loss_rescaled  time_train_s_rescaled  time_infer_s_rescaled      rank  rank=1_count  rank=2_count  rank=3_count  rank>3_count  error_count
0      AutoGluon_holdout_1h_2021_06_08_holdout                                             76                                             64                                              0                                           0.092548                     3836.912143    391.401929     65.761429       0.457143               1.003535               1.072343  1.457143            76            64             0             0            0
1  AutoGluon_bestquality_1h_2021_06_08_holdout                                              0                                              0                                            140                                           0.000000                     3904.792857    393.050993     86.070000       0.542857               1.021506               1.956817  1.542857            64            76             0             0            0

```

Future PR's can expand this functionality by enabling `use_bag_holdout` option to be refit into the standard stack, which while taking twice as long to train, should produce stronger results in general.

**Comment by szha** [comment]
Job PR-1105-7 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1105/7/index.html

**Comment by szha** [comment]
Job PR-1105-10 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1105/10/index.html

File: core/src/autogluon/core/utils/utils.py [hunk 1]
```diff
@@ -273,26 +300,82 @@ def get_pred_from_proba(y_pred_proba, problem_type=BINARY):
...
+        y_train = y_train.append(y.loc[rare_indices])
+
+    if problem_type in [BINARY, MULTICLASS]:
+        class_counts_dict_orig = y.value_counts().to_dict()
+        class_counts_dict = y_train.value_counts().to_dict()
+        class_counts_dict_test = y_test.value_counts().to_dict()
+
+        indices_to_move = []
+        random_state = random.getstate()
+        random.seed(random_state)
```
Comment (gradientsky) [inline core/src/autogluon/core/utils/utils.py:359]: Why is this needed? It looks like noop.

**Review by gradientsky — CHANGES_REQUESTED** [review]

File: core/src/autogluon/core/utils/utils.py (hunk 1 above)
Comment (Innixma) [reply core/src/autogluon/core/utils/utils.py:359]: Good catch! This was a bug, it is supposed to apply the random_state param and then reset to the outer random_state context.

Comment (Innixma) [reply core/src/autogluon/core/utils/utils.py:359]: Fixed

File: core/src/autogluon/core/utils/utils.py [hunk 2]
```diff
@@ -240,13 +259,21 @@ def augment_rare_classes(X, label, threshold):
             new_df = new_df.append(clss_df.copy())
         aug_df = aug_df.append(new_df.copy())
 
+    # Ensure new samples generated via augmentation have unique indices
+    aug_df = aug_df.reset_index(drop=True)
+    aug_df_len = len(aug_df)
+    X_index_aug_start = max(list(X.index)) + 1
```
[the remark this answers is not in the document]
Comment (Innixma) [reply core/src/autogluon/core/utils/utils.py:265]: Added

File: core/src/autogluon/core/utils/utils.py [hunk 3]
```diff
@@ -47,7 +49,24 @@ def generate_kfold(X, y=None, n_splits=5, random_state=0, stratified=False, n_re
...
+                out = [[train_index, test_index] for train_index, test_index in kf.split(X, y)]
+            except:
+                y_dummy = pd.concat([y, pd.Series([-1] * n_splits)], ignore_index=True)
+                X_dummy = pd.concat([X, X.head(n_splits)], ignore_index=True)
+                invalid_index = set(list(y_dummy.tail(n_splits).index))
+                out = [[train_index, test_index] for train_index, test_index in kf.split(X_dummy, y_dummy)]
+                len_out = len(out)
+                for i in range(len_out):
+                    train_index = out[i][0]
+                    test_index = out[i][1]
```
[the remark this answers is not in the document]
Comment (Innixma) [reply core/src/autogluon/core/utils/utils.py:66]: Added

**Review by gradientsky — APPROVED** [review]

**Comment by szha** [comment]
Job PR-1105-11 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1105/11/index.html
